### ISSUE TITLE
Fail to download video

### ISSUE THREAD

**Issue by smediomike** [aws-amplify/amplify-swift#2748]
### Describe the bug

when I download 1000 pictures+20 videos from s3. it sometimes fail to download video file(very few happen on picture) ; Amplify 1.0/1.28/2.30.1 all happen such issues. this issue happen on iPhone 7 Plus(14.5)/iPhoneX(16.3)/iPhone12(16.3) test devices all.

        let fileUrl = URL(fileURLWithPath: local_path)
        let soption : StorageDownloadFileRequest.Options =  StorageDownloadFileRequest.Options(accessLevel: StorageAccessLevel.private)
        
        let storageOperation = Amplify.Storage.downloadFile(
            key: key,
            local: fileUrl,
            options: soption,
            progressListener: { progress in
                size_progress(progress.completedUnitCount)
            }, resultListener: {
                // result handle
         }
}

### Steps To Reproduce

```swift
Steps to reproduce the behavior:
1. get s3 key + metadata information from datastore
2. download picture + video content according to media type + s3 key
3. after finished download pictures, and then may fail to download video file sometimes.
```

### Expected behavior

download all request files successfully in one time.

### Amplify Framework Version

2.30.1

### Amplify Categories

Storage

### Dependency manager

Cocoapods

### Swift version

5.0

### CLI version

10.4.1

### Xcode version

Xcode 14.2

### Relevant log output

```shell
when happen this issue, will print the below log:
unknown
Swift DwonloadFile Failed: Unexpected error occurred with message: Domain: [NSCocoaErrorDomain
Code: [4
LocalizedDescription: [The file doesn’t exist.
LocalizedFailureReason: [no error
LocalizedRecoverySuggestion: [. This should not happen. There is a possibility that there is a bug if this error persists. Please take a look at https://github.com/aws-amplify/amplify-ios/issues to see if there are any existing issues that match your scenario, and file an issue with the details of the bug if there isn't. Issue encountered at:
file: /Users/wushujiang/p4/temp_p4/IOS/BRANCH/CloudBackup/CloudBackup/Pods/Amplify/Amplify/Categories/Storage/Error/StorageError.swift
function: recoverySuggestion
line: 63. 7
```

### Is this a regression?

No

### Device

iPhoneX(16.3)/iPhone12(16.3)/iPhone7 plus(14.5)

### iOS Version

16.3/14.5

**Comment by jcjimenez** [comment]
Hi, @ smediomike, I haven't confirmed, but it looks like the library may not be guarding against non-existent directories (if so, I will submit a fix for this).

However, to save you time, I figured I'd ask: is it possible that one of the intermediate directories within `local_path` doesn't exist? For example, the code below will throw the same error you're seeing:

```
        let directory = NSTemporaryDirectory().appending("/\(UUID().uuidString)")
        let file = directory.appending("/\(UUID().uuidString)")
        let data = Data(UUID().uuidString.utf8)
        try data.write(to: URL(fileURLWithPath: file))
```

And adding this just before the call to `Data.write(to:)`, makes it work:

```
        try FileManager.default.createDirectory(atPath: directory, withIntermediateDirectories: true)
```

**Comment by smediomike** [comment]
Hello Jimenez,


I changed possible place as you meant to promise the folder exist, I diffcult to reproduce this issue, thanks for your help. if have any further information, i will send to you again.


Best Regards,

mike.wu

在 2023年2月9日 04:00，JC ***@***.***> 写道：


Hi, @ smediomike, I haven't confirmed, but it looks like the library may not be guarding against non-existent directories (if so, I will submit a fix for this).

However, to save you time, I figured I'd ask: is it possible that one of the intermediate directories within local_path doesn't exist? For example, the code below will throw the same error you're seeing:

        let directory = NSTemporaryDirectory().appending("/\(UUID().uuidString)")
        let file = directory.appending("/\(UUID().uuidString)")
        let data = Data(UUID().uuidString.utf8)
        try data.write(to: URL(fileURLWithPath: file))


And adding this just before the call to Data.write(to:), makes it work:

        try FileManager.default.createDirectory(atPath: directory, withIntermediateDirectories: true)
